File: frontend/src/types/agent.ts
import type { AgentType, TaskStatus } from "./contracts";
import type { KnowledgeNode } from "./course";
import type { LearningPath } from "./learningPath";
import type { StudentProfile } from "./profile";
import type { LearningRecord } from "./report";
import type { RetrievedDocument } from "./resource";

export interface ChatSession {
  id: string;
  userId: string;
  courseId?: string;
  nodeId?: string;
  title: string;
  sessionType: "profile" | "qa" | "resource" | "practice";
  createdAt: string;
  updatedAt: string;
}

export interface ChatMessage {
  id: string;
  sessionId: string;
  userId: string;
  role: "user" | "assistant" | "system" | "agent";
  content: string;
  contentType: "text" | "markdown" | "json";
  agentType?: AgentType;
  createdAt: string;
}

export interface ChatRequest {
  userId: string;
  sessionId?: string;
  courseId?: string;
  nodeId?: string;
  message: string;
  useRag: boolean;
  useProfile: boolean;
}

export interface ChatResult {
  sessionId: string;
  messageId: string;
  answer: string;
  usedAgentTypes: AgentType[];
  retrievedDocuments?: RetrievedDocument[];
}

export interface ChatStreamEvent {
  sessionId: string;
  eventType: "start" | "chunk" | "agent_step" | "done" | "error";
  contentChunk?: string;
  agentType?: AgentType;
  errorMessage?: string;
}

export interface AgentContext {
  profile?: StudentProfile;
  currentNode?: KnowledgeNode;
  learningPath?: LearningPath;
  recentRecords?: LearningRecord[];
  retrievedDocuments?: RetrievedDocument[];
}

export interface AgentRunRequest {
  userId: string;
  courseId?: string;
  nodeId?: string;
  agentType: AgentType;
  input: Record<string, any>;
  context?: AgentContext;
}

export interface AgentRunResult {
  taskId: string;
  agentType: AgentType;
  status: TaskStatus;
  output: Record<string, any>;
  errorMessage?: string;
}

export interface MultiAgentWorkflowRequest {
  userId: string;
  courseId: string;
  nodeId?: string;
  workflowType:
    | "profile_build"
    | "path_plan"
    | "resource_generate"
    | "qa"
    | "practice_review"
    | "report_generate";
  input: Record<string, any>;
}

export interface MultiAgentWorkflowResult {
  taskId: string;
  workflowType: string;
  status: TaskStatus;
  steps: AgentRunResult[];
  finalOutput: Record<string, any>;
}

export interface AgentTaskEvent {
  taskId: string;
  agentType: AgentType;
  eventType: "start" | "thinking" | "tool_call" | "result" | "error" | "done";
  message?: string;
  payload?: Record<string, any>;
  createdAt: string;
}

File: backend/app/schemas/agent.py
from typing import Any, Literal

from pydantic import Field

from app.schemas.course import KnowledgeNode
from app.schemas.learning_path import LearningPath
from app.schemas.profile import StudentProfile
from app.schemas.report import LearningRecord
from app.schemas.resource import RetrievedDocument
from app.schemas.common import AgentType, ContractModel, TaskStatus


class ChatSession(ContractModel):
    id: str
    user_id: str
    course_id: str | None = None
    node_id: str | None = None
    title: str
    session_type: Literal["profile", "qa", "resource", "practice"]
    created_at: str
    updated_at: str


class ChatMessage(ContractModel):
    id: str
    session_id: str
    user_id: str
    role: Literal["user", "assistant", "system", "agent"]
    content: str
    content_type: Literal["text", "markdown", "json"]
    agent_type: AgentType | None = None
    created_at: str


class ChatRequest(ContractModel):
    user_id: str
    session_id: str | None = None
    course_id: str | None = None
    node_id: str | None = None
    message: str
    use_rag: bool
    use_profile: bool


class ChatResult(ContractModel):
    session_id: str
    message_id: str
    answer: str
    used_agent_types: list[AgentType] = Field(default_factory=list)
    retrieved_documents: list[Any] | None = None


class ChatStreamEvent(ContractModel):
    session_id: str
    event_type: Literal["start", "chunk", "agent_step", "done", "error"]
    content_chunk: str | None = None
    agent_type: AgentType | None = None
    error_message: str | None = None


class AgentContext(ContractModel):
    profile: StudentProfile | None = None
    current_node: KnowledgeNode | None = None
    learning_path: LearningPath | None = None
    recent_records: list[LearningRecord] | None = None
    retrieved_documents: list[RetrievedDocument] | None = None


class AgentRunRequest(ContractModel):
    user_id: str
    course_id: str | None = None
    node_id: str | None = None
    agent_type: AgentType
    input: dict[str, Any]
    context: AgentContext | None = None


class AgentRunResult(ContractModel):
    task_id: str
    agent_type: AgentType
    status: TaskStatus
    output: dict[str, Any]
    error_message: str | None = None


class MultiAgentWorkflowRequest(ContractModel):
    user_id: str
    course_id: str
    node_id: str | None = None
    workflow_type: Literal["profile_build", "path_plan", "resource_generate", "qa", "practice_review", "report_generate"]
    input: dict[str, Any]


class MultiAgentWorkflowResult(ContractModel):
    task_id: str
    workflow_type: str
    status: TaskStatus
    steps: list[AgentRunResult] = Field(default_factory=list)
    final_output: dict[str, Any]


class AgentTaskEvent(ContractModel):
    task_id: str
    agent_type: AgentType
    event_type: Literal["start", "thinking", "tool_call", "result", "error", "done"]
    message: str | None = None
    payload: dict[str, Any] | None = None
    created_at: str
